fog/types/src/common.rs

```diff
@@ -61,6 +61,51 @@ impl BlockRange {
     pub fn is_empty(&self) -> bool {
         self.len() == 0
     }
+
+    /// Merge the first contiguous ranges into a range
+    ///
+    /// The ranges should be sorted to produce consistent results.
+    /// They are not sorted here in order to properly indicate the index of the
+    /// last range merged.
+    ///
+    /// ```
+    /// use mc_fog_types::common::BlockRange;
+    ///
+    /// let ranges = vec![BlockRange::new(0, 5), BlockRange::new(3, 8), BlockRange::new(8, 10), BlockRange::new(15, 16)];
+    /// let (index, merged_range) = BlockRange::merge_from_ranges(ranges.iter()).unwrap();
+    /// assert_eq!(index, 2);
+    /// assert_eq!(merged_range, BlockRange::new(0, 10));
+    ///
+    /// let ranges = vec![BlockRange::new(3, 8), BlockRange::new(0, 5), BlockRange::new(8, 10), BlockRange::new(15, 16)];
+    /// let (index, merged_range) = BlockRange::merge_from_ranges(ranges.iter()).unwrap();
+    /// assert_eq!(index, 0);
+    /// assert_eq!(merged_range, BlockRange::new(3, 8));
+    /// ```
+    ///
+    /// Returns the index in `ranges` of the last block range merged, and the
+    /// merged range
+    pub fn merge_from_ranges<'a>(
+        ranges: impl IntoIterator<Item = &'a BlockRange>,
+    ) -> Option<(usize, BlockRange)> {
```

Comment:
My thought is that the caller is going to want to know which was the last index used for the range and that will be the one whose values of max_tx_count and what not are used